### PR TITLE
Support tracing base torch_function impl

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111765
* #111737
* __->__ #111731
* #111730

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -478,6 +479,29 @@ def make_const_size_variable(x, **options):
...
+
+            # [Note: __torch_function__] coerce this tensor variable into a TensorWithTFOverrideVariable
+            # in eager, this is just a type change. This isn't sound if a __torch_function__ tensor subclass
+            # defines a constructor, but if only a __torch_function__ impl is defined, this is okay to call.
+            # It is up to the user whether this is correct behavior or not.
+            py_cls = args[0].as_python_constant()
+            torch_fn = VariableBuilder(
+                tx,
+                AttrSource(
+                    AttrSource(args[0].source, "__torch_function__"), "__func__"
```
Comment (jansel) [inline torch/_dynamo/variables/tensor.py:498]: Need to check `args[0].source` is set.

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -245,6 +241,16 @@ def call_function(
...
             ).call_function(tx, args, kwargs)
+        if self.value is torch.overrides.get_default_nowrap_functions:
+            # [Note: __torch_function__] we return empty here because we restrict
+            # the set of functions that we trace __torch_function__ on to
+            # functions outside of the actual set. Implementing this properly will require implementing
+            # some variable types to track and compare tensor getset descriptors
+            from .builder import SourcelessBuilder
+
+            return SourcelessBuilder()(
+                tx, torch.overrides.get_default_nowrap_functions()
```
Comment (jansel) [inline torch/_dynamo/variables/torch.py:252]: Do we need a guard for this?  Or is it a constant?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/variables/torch.py:252]: It's a constant

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**